Upgrade Nudge: Fix Styling by using `@wordpress/editor`'s `Warning` component

ockham opened the pull request:
Fixes #12946

#### Changes proposed in this Pull Request:

* Upgrade Nudge: Fix Styling by using `@wordpress/editor`'s [`Warning`](https://github.com/WordPress/gutenberg/blob/21e2dce2b287347e21c3e0618bc8fa8c0c1ef5ca/packages/block-editor/src/components/warning/index.js) component 

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Fix styling of the recently introduced Upgrade Nudge

#### Screenshots

![image](https://user-images.githubusercontent.com/96308/60708281-87139e80-9f2f-11e9-8302-be57580e76a4.png)

![image](https://user-images.githubusercontent.com/96308/60708307-9266ca00-9f2f-11e9-85f9-e01bd14b8a2d.png)

#### Testing instructions:

- Switch to this branch locally, build Jetpack (`yarn build`), and start Docker (`yarn docker:up`)
- Now add the following line to any file in `docker/mu-plugins`, e.g. a newly created `0-blocks-upgrade.php` (needs to start with `<?php`): `define ( 'JETPACK_SHOW_BLOCK_UPGRADE_NUDGE', true );`
- Make sure you're connected to WP.com, and on a free plan
- Start a new post, and insert the 'Simple Payments' block
- It should come with an Upgrade Nudge on top of it, which should look as depicted in the screenshots.

#### Proposed changelog entry for your changes:

(Covered by whatever we add for #12823)

ockham committed:
Upgrade Nudge: Use Gutenberg Font, Line Height

ockham committed:
Remove hue

ockham committed:
Use moar Gutenberg styling

ockham committed:
Use Warning component from @wordpress/editor

ockham committed:
Add back Gridicon

ockham committed:
Add back Gridicon. :heart: Flexbox

ockham committed:
Add border

ockham committed:
Change second line of copy

ockham committed:
Only add border if not selected

ockham committed:
Bottom margin/padding doesn't work as I hoped

simison commented on extensions/shared/wrap-paid-block/style.scss:
```diff
@@ -1,8 +1,10 @@
-.jetpack-paid-block__disabled {
-	opacity: 0.5;
-	padding: 10px;
+.block-editor-block-list__block:not( .is-selected ) .jetpack-paid-block__wrapper {
```
Instead of referencing core class-names (which can be fragile), you could pass `.is-selected` class for `UpgradeNudge` in the HOC since you can see it in `props.isSelected`. Not a blocker and up to you how you see best.

ockham replied:
:+1: 5f13221

simison reviewed (dismissed):
Works well :thumbsup:

Ok to merge already, but I wanted to point out these things that you could either fix in this PR or in follow-ups:

- There's a little off-border leak visible here:
    <img width="137" alt="Screenshot 2019-07-05 at 12 35 53" src="https://user-images.githubusercontent.com/87168/60713806-6729ae00-9f22-11e9-8b03-1cfe007e3cbb.png">

- I'd prefer to have 14px empty space between the block content and the banner:

    <img width="343" alt="Screenshot 2019-07-05 at 12 43 52" src="https://user-images.githubusercontent.com/87168/60713950-abb54980-9f22-11e9-8ffc-35d8006fdba8.png">

    Might just need `margin-bottom: 0`:
    <img width="289" alt="Screenshot 2019-07-05 at 12 43 48" src="https://user-images.githubusercontent.com/87168/60713951-abb54980-9f22-11e9-8a1a-8f3ac2b14588.png">



- I'd expect upgrade banner to get a blue hue when multiple blocks are selected:

    <img width="817" alt="image" src="https://user-images.githubusercontent.com/87168/60713769-57aa6500-9f22-11e9-9223-7b2ad59eb5df.png">

    — might be just a matter of correct z-index or `background:transparent`?

ockham commented:
Also noting that the nudge doesn't look great on WP.com:

![image](https://user-images.githubusercontent.com/96308/60714449-e926d080-9f3c-11e9-8b1f-ecb8063dfc9d.png)

This is apparently because of `margin` styling for the `.editor-styles-wrapper p` selector:

![image](https://user-images.githubusercontent.com/96308/60714734-7702bb80-9f3d-11e9-9175-47e517e851ea.png)

Looks like overrides on top of overrides :sob:

If I disable all 4 `margin` relevant checkboxes in the element inspector as seen above, the spacing looks fine:

![image](https://user-images.githubusercontent.com/96308/60714847-a285a600-9f3d-11e9-82f6-35c817633036.png)

ockham committed:
Space below Upgrade Nudge

simison commented:
On small screens left border is cut out:

<img width="376" alt="image" src="https://user-images.githubusercontent.com/87168/60716329-c50cc480-9f27-11e9-92d3-cfd53db164d0.png">

ockham committed:
Fix off-border leak

simison commented on extensions/shared/wrap-paid-block/style.scss:
```diff
@@ -1,8 +1,10 @@
+	margin-left: -14px;
+	margin-right: -14px;
+	padding-left: calc( 14px - #{$border-width} );
+	padding-right: calc( 14px - #{$border-width} );
```
14px might be good candidate for a variable: https://github.com/WordPress/gutenberg/blob/9052d4c26fff89907d7e7abec745a3d0079dd178/assets/stylesheets/_variables.scss#L57

ockham replied:
6ecaace

ockham committed:
Fix WP.com styling

ockham committed:
Use block-padding var

ockham committed:
Rearrange stylesheet a bit

ockham committed:
Better is-selected handling

ockham commented:
> Works well 
> 
> Ok to merge already, but I wanted to point out these things that you could either fix in this PR or in follow-ups:
> 
> * There's a little off-border leak visible here:
>   <img alt="Screenshot 2019-07-05 at 12 35 53" width="137" src="https://user-images.githubusercontent.com/87168/60713806-6729ae00-9f22-11e9-8b03-1cfe007e3cbb.png">

Fixed by e4640ff

> * I'd prefer to have 14px empty space between the block content and the banner:
>     <img alt="Screenshot 2019-07-05 at 12 43 52" width="343" src="https://user-images.githubusercontent.com/87168/60713950-abb54980-9f22-11e9-8ffc-35d8006fdba8.png">
>   Might just need `margin-bottom: 0`:

Fixed by 0bd7b58

>   <img alt="Screenshot 2019-07-05 at 12 43 48" width="289" src="https://user-images.githubusercontent.com/87168/60713951-abb54980-9f22-11e9-8a1a-8f3ac2b14588.png">
> * I'd expect upgrade banner to get a blue hue when multiple blocks are selected:
>     <img alt="image" width="817" src="https://user-images.githubusercontent.com/87168/60713769-57aa6500-9f22-11e9-9223-7b2ad59eb5df.png">
>   — might be just a matter of correct z-index or `background:transparent`?

Ideally, yeah. However, I've seen a bunch of code and/or styling in core (e.g. in the `Warning` component that this PR uses) that says something like 'make sure it works with multi-block selection' (or something like that), so it might be a tad more involved :grimacing:

ockham committed:
Remove obsolete flex property

ockham commented:
> Also noting that the nudge doesn't look great on WP.com:
> 
> ![image](https://user-images.githubusercontent.com/96308/60714449-e926d080-9f3c-11e9-8b1f-ecb8063dfc9d.png)
> 
> This is apparently because of `margin` styling for the `.editor-styles-wrapper p` selector:
> 
> ![image](https://user-images.githubusercontent.com/96308/60714734-7702bb80-9f3d-11e9-9175-47e517e851ea.png)
> 
> Looks like overrides on top of overrides 
> 
> If I disable all 4 `margin` relevant checkboxes in the element inspector as seen above, the spacing looks fine:
> 
> ![image](https://user-images.githubusercontent.com/96308/60714847-a285a600-9f3d-11e9-82f6-35c817633036.png)

Fixed by e30a44e

simison reviewed (dismissed):
Looking very solid already!

Only things to follow up is I think fine-tuning padding/margins, e.g. there's this on hover:

<img width="558" alt="Screenshot 2019-07-05 at 15 56 01" src="https://user-images.githubusercontent.com/87168/60723836-6fdbad80-9f3d-11e9-9f1f-1fff88b69261.png">

And the banner seems to jump ~1-2px when selecting the block so it doesn't feel as smooth as it does for "missing block" and similar core alerts.

Multi-selection block styles we can improve in other PR, especially if it's complicated.

Good to merge 👍

simison commented:
Not needed for this PR but just noting that we might want to add this colorful border to align visually more with Calypso:

<img width="731" alt="Screenshot 2019-07-08 at 09 45 15" src="https://user-images.githubusercontent.com/87168/60788388-4735fc80-a165-11e9-955b-ec9a85005fdc.png">

We just need to figure out first if it's meant only for banners that are clickable and not for banners that have CTA button separately.

It might also just look silly with the thick grey border when hovering to block, let's see...

mtias commented:
Can you try with the button vertically centered?

simison committed:
Center align CTA button

simison committed:
Stop icon from shrinking (+abc'ify)

simison committed:
Upgrade copy

See https://github.com/Automattic/jetpack/issues/12946#issuecomment-508745047

simison commented:
>Can you try with the button vertically centered?

Fixed in 0da343ac3

Also upgraded the copy with a suggestion from https://github.com/Automattic/jetpack/issues/12946#issuecomment-508745047:

> This block is available under the Premium Plan.
> It will be hidden from site visitors until you upgrade.

zinigor reviewed (approved):
Works well for me!
